### DISCUSSION TITLE
Confirm running a command for each file in the quickfix list ?

### DISCUSSION THREAD

**Discussion by adam-devel** [neovim/neovim#31429] [Q&A]
After using `:grep`, `:vimgrep` or some other command to populate the quickfix list with matches, one could run `:cdo !command` to run a command on all the matched files.

Is there a way to prompt the user to confirm running the command for each file ?
The reason I want to do this is to avoid running the command on false positives.

One alternative may be allowing the user to go over the files one by one and choosing whether to keep or remove the file from the list. thereby pruning it from false positives before running the command.

Either way, It becomes real tedious real fast if I have to manually perform a complex operation on every single file in the list just to confirm or skip, the delays add up in a long list. the confirmation has to be done in a streamlined way.

I guess I could offload this work to a script that prompts me before each operation, but that's a last resort. and not in the spirit of the quick fix list (you may disagree but it's out of topic for this question).

If neovim doesn't support this, let me know so I can propose it as feature request. I would like to avoid using plugins for this because I think it should be part of `:cdo`, maybe a flag supplied to it or an alternative command that asks for confirmation, this would complete the functionality of `:cdo`.

**Comment by zeertzjq** [comment]
You can define a user command that prompts for confirmation before executing a command:
```vim
:command! -nargs=+ -complete=command Confirm if confirm('Execute "' .. expandcmd(<q-args>) .. '"?', "&Yes\n&No") == 1 | execute <q-args> | endif
```
Then you can use for example:
```vim
:cdo Confirm !cat %
```
I don't think this should be built-in, as different people may have different idea on what the confirmation should look like.

**Reply by adam-devel** [reply]
That is not convenient.
The quickfix list is a list. the only reason you would use a list is to group things for convenience, you can easily traverse the list with :cnext or operate on the whole list with :cdo. the whole point is operating on multiple items conveniently.

Substitution (an operation on multiple items/matches) provides the 'c' flag by default, which asks for confirmation before operating. this makes sense, it's less destructive and it's often the case that you want to avoid false positives. it comes with the territory whenever you are operating on multiple items

> different people may have different idea on what the confirmation should look like

that never stopped neovim from having a reasonable default. we have the precedence of the 'c' flag. a simple Y/n prompt with an additional option to quit would make sense.

people want and do change all aspects of neovim, yet neovim exists as an editor with it's own choices. this isn't a library waiting to be consumed the way the user wants to, it's an editor with reasonable defaults that can be changed

**Reply by zeertzjq** [reply]
If adding a single word isn't convenient enough, then there is no convenient way to do this. `:cdo` is an Ex command and its `!` modifier is already taken, while its argument is taken as a full command, so there is no place to add a confirmation flag other than adding a word before the argument.

**Reply by zeertzjq** [reply]
Also keep in mind that consistency is also an important factor. There are 8 `:*do` commands:
- `:argdo`
- `:bufdo`
- `:cdo`
- `:cdfo`
- `:ldo`
- `:lfdo`
- `:tabdo`
- `:windo`

Adding a confirmation prompt for only one of them is inconsistent. Adding confirmation prompts for all of them is even harder to design, as the relevant information is different for different `:*do` commands.

**Reply by adam-devel** [reply]
> If adding a single word isn't convenient

the following isn't a single word:
```vim
:command! -nargs=+ -complete=command Confirm if confirm('Execute "' .. expandcmd(<q-args>) .. '"?', "&Yes\n&No") == 1 | execute <q-args> | endif
```

> there is no place to add a confirmation flag

that's an implementation detail that can be discussed. notice how there's `:ldo` as well as `:lfdo`. the only reason `:lfdo` exists is because adding a flag doesn't work for the reasons you outlined. this is a separate issue.

> Adding a confirmation prompt for only one of them is inconsistent

yes, if it gets added to one it should get added to all of them. if the code is well organized it shouldn't be as hard as re-implementing the same feature for every command. this isn't the main point really. i am arguing that the feature is essential when doing potentially destructive operations on lists of things. once that's evident, the details and roadblocks can discussed

**Reply by zeertzjq** [reply]
Thinking about this again, this may indeed require builtin syntax, as different `:*do` commands require different confirmation messages, which is hard to achieve with a user command.

Please see https://github.com/vim/vim/issues/16176

**Reply by adam-devel** [reply]
Thank you for filing the issue. should i close this as resolved ? or should this wait until the issue is resolved ?

**Reply by ars-celare** [reply]
This is essential.

>Substitution (an operation on multiple items/matches) provides the 'c' flag by default, which asks for confirmation before operating. this makes sense, it's less destructive and it's often the case that you want to avoid false positives. it comes with the territory whenever you are operating on multiple items

Absolutely. The fact you can't easily do this with any other commands is a bizarre and glaring omission (just so strange! The quickfix list gets hyped up as this awesome tool, a killer app for vim... but it's just so bizarre and disappointing you can't do the obvious thing with it, at least not safely...)

We need editors nowadays to be competitive with AI, so this fix is of the utmost importance.
